applications/nrf_desktop/src/modules/ble_discovery.c

```diff
@@ -184,45 +218,78 @@ static void dis_read_completed(struct bt_gatt_dm *dm, void *context)
 
 	rp.single.handle = attr->handle + 1;
 
+	err = bt_gatt_read(bt_gatt_dm_conn_get(dm), &rp);
+	if (err) {
+		LOG_ERR("GATT read problem (err:%d)", err);
+		peer_unpair();
+	}
+
 	err = bt_gatt_dm_data_release(dm);
 	if (err) {
 		goto error;
 	}
 
-	bt_gatt_read(discovering_peer_conn, &rp);
 	return;
 error:
 	LOG_ERR("Discovery data release failed (err:%d)", err);
 	module_set_state(MODULE_STATE_ERROR);
 }
 
-static void dis_not_found(struct bt_conn *conn, void *context)
+static void start_discovery(struct bt_conn *conn,
+			    const struct bt_uuid *svc_uuid)
 {
-	LOG_ERR("DIS not found");
-	__ASSERT_NO_MSG(discovering_peer_conn == conn);
-	peer_unpair();
+	static const struct bt_gatt_dm_cb dis_discovery_cb = {
+		.completed = discovery_completed,
+		.service_not_found = service_not_found,
+		.error_found = discovery_error,
+	};
+
+	int err = bt_gatt_dm_start(conn, svc_uuid, &dis_discovery_cb, NULL);
+
+	if (err) {
+		LOG_ERR("Cannot start discovery (err:%d)", err);
```

Comment:
We are not just logging the problem. We also unpair and disconnect peer. I think there is nothing more we could do here. In addition this should not happen with current implementation.

Reply:
If there is an execution path the it has to be safe. What I ment, are you sure that in case of error is the app able to connect to other device and do GATT discovery?

Reply:
Ok, I will add assert that error is not -EALREADY nor -EINVAL (to catch it for debug configurations). And in general for any error it will remove the peer.

Reply:
ok